### ISSUE TITLE
Ivar.fill crash investigation

### ISSUE THREAD

**Issue by ghost-not-in-the-shell** [MinaProtocol/mina#7463]
# Things that look a little suspicious:
https://github.com/MinaProtocol/mina/blob/88699ca3ec4ff273bfe7c1e798447349f5029ea8/src/lib/distributed_dsl/distributed_dsl.ml#L196
This looks suspicious (I stared at the code for a while, but didn't find a way to defend it), but I think this file is not being used.

---

https://github.com/MinaProtocol/mina/blob/88699ca3ec4ff273bfe7c1e798447349f5029ea8/src/lib/mina_net2/mina_net2.ml#L60
https://github.com/MinaProtocol/mina/blob/88699ca3ec4ff273bfe7c1e798447349f5029ea8/src/lib/mina_net2/mina_net2.ml#L63
The logic here is complicated. So it could be suspicious.

---

https://github.com/MinaProtocol/mina/blob/88699ca3ec4ff273bfe7c1e798447349f5029ea8/src/lib/transition_handler/processor.ml#L318
The old catchup implementation only writes to catchup pipe once, so does the new implementation. But since the crashes happens when catchup happens so it is also suspicious.

---

https://github.com/MinaProtocol/mina/blob/88699ca3ec4ff273bfe7c1e798447349f5029ea8/src/lib/transition_frontier_controller/transition_frontier_controller.ml#L72
This is suspicious because the `Ivar.fill` crashes happens around when `Transition_frontier_controller` closes.

---

https://github.com/MinaProtocol/mina/blob/88699ca3ec4ff273bfe7c1e798447349f5029ea8/src/lib/transition_frontier/extensions/transition_registry.ml#L16
The logic here is again complicated. So it could be suspicious.

---

# Things that look benign:
https://github.com/MinaProtocol/mina/blob/88699ca3ec4ff273bfe7c1e798447349f5029ea8/src/lib/block_producer/block_producer.ml#L34
The `cancel` function is only used in the `dispatch` function below. The usage of `Ivar.fill` should be safe because each time `cancel` gets called, a new ivar would be created in place of the old ivar.

---

https://github.com/MinaProtocol/mina/blob/88699ca3ec4ff273bfe7c1e798447349f5029ea8/src/lib/block_producer/block_producer.ml#L62
The usage of `Ivar.fill` follows the pattern `Deferred.create (fun ivar -> ... ; Ivar.fill ivar ())`, so it should be obviously safe.

---

https://github.com/MinaProtocol/mina/blob/88699ca3ec4ff273bfe7c1e798447349f5029ea8/src/lib/cache_lib/impl.ml#L132
`mark_failed/success` function is only used in the `invalidate_with_failure` function. The usage of `Ivar.fill` is protected by assertions inside those two functions.

---

https://github.com/MinaProtocol/mina/blob/88699ca3ec4ff273bfe7c1e798447349f5029ea8/src/lib/child_processes/child_processes.ml#L316
`terminated_ivar` is created when the process is created, which means this ivar goes with each individual child process. Only when the process exits or crashes, `terminated_ivar` would be filled. So it should be safe.

---

https://github.com/MinaProtocol/mina/blob/88699ca3ec4ff273bfe7c1e798447349f5029ea8/src/lib/consensus/proof_of_stake.ml#L2751
The usage of `Ivar.fill` follows the pattern `Deferred.create (fun ivar -> ... ; Ivar.fill ivar ())`, so it should be obviously safe.

---

https://github.com/MinaProtocol/mina/blob/88699ca3ec4ff273bfe7c1e798447349f5029ea8/src/lib/integration_test_cloud_engine/stack_driver_log_engine.ml#L638
This should be safe because it tries to fill in a locally created ivar. Plus this file should not be used by the daemon.

---

https://github.com/MinaProtocol/mina/blob/88699ca3ec4ff273bfe7c1e798447349f5029ea8/src/lib/integration_test_cloud_engine/stack_driver_log_engine.ml#L705
https://github.com/MinaProtocol/mina/blob/88699ca3ec4ff273bfe7c1e798447349f5029ea8/src/lib/integration_test_cloud_engine/stack_driver_log_engine.ml#L740
The above two usage should be safe because they are already been protected by `Ivar.is_empty`/`Ivar.is_full` respectively. Plus this file should not be used by the daemon.

---

https://github.com/MinaProtocol/mina/blob/88699ca3ec4ff273bfe7c1e798447349f5029ea8/src/lib/mina_lib/mina_lib.ml#L178
This should be fine because this code path would only be called once when the process exits or crashes.

---

https://github.com/MinaProtocol/mina/blob/88699ca3ec4ff273bfe7c1e798447349f5029ea8/src/lib/mina_lib/mina_lib.ml#L226
This should be fine because `Snark_worker.start` is only called once in `Mina_lib.start`.

---

https://github.com/MinaProtocol/mina/blob/88699ca3ec4ff273bfe7c1e798447349f5029ea8/src/lib/mina_lib/mina_lib.ml#L714
This should be fine because it tries to fill in a locally created ivar.

---

https://github.com/MinaProtocol/mina/blob/88699ca3ec4ff273bfe7c1e798447349f5029ea8/src/lib/mina_net2/mina_net2.ml#L762
There's already a comment saying this `fill` is safe:)

---

https://github.com/MinaProtocol/mina/blob/88699ca3ec4ff273bfe7c1e798447349f5029ea8/src/lib/network_pool/batcher.ml#L71
https://github.com/MinaProtocol/mina/blob/88699ca3ec4ff273bfe7c1e798447349f5029ea8/src/lib/network_pool/batcher.ml#L74
https://github.com/MinaProtocol/mina/blob/88699ca3ec4ff273bfe7c1e798447349f5029ea8/src/lib/network_pool/batcher.ml#L85
This recursive function should be fine because it filtered out the filled element in the list.

---

https://github.com/MinaProtocol/mina/blob/88699ca3ec4ff273bfe7c1e798447349f5029ea8/src/lib/pipe_lib/broadcast_pipe.ml#L35
This should be fine because it fills in a locally created ivar.

---

https://github.com/MinaProtocol/mina/blob/88699ca3ec4ff273bfe7c1e798447349f5029ea8/src/lib/syncable_ledger/syncable_ledger.ml#L460
This should be fine, because it only gets called once the sync is done.

---

https://github.com/MinaProtocol/mina/blob/88699ca3ec4ff273bfe7c1e798447349f5029ea8/src/lib/timeout_lib/timeout_lib.ml#L85
This should be fine, because it follows the pattern `Deferred.create (fun ivar -> ... ; Ivar.fill ivar ())`.

**Comment by deepthiskumar** [comment]
No new reports of this
